code/modules/awaymissions/mission_code/ruins/watcher_grave.dm

```diff
@@ -0,0 +1,256 @@
+
+/// Egg which hatches into a helpful pet. Or you can eat it if you want.
+/obj/item/reagent_containers/food/snacks/egg/watcher
+	name = "watcher egg"
+	desc = "A lonely egg still pulsing with life, somehow untouched by the corruption of the Necropolis."
+	icon_state = "egg_watcher"
+	tastes = list("ocular fluid" = 6, "loneliness" = 1)
+	antable = FALSE
+	/// How far have we moved?
+	var/steps_travelled = 0
+	/// How far should we travel to hatch?
+	var/steps_to_hatch = 600
+	/// Datum used to measure our steps
+	var/datum/movement_detector/pedometer
+
+/obj/item/reagent_containers/food/snacks/egg/watcher/Initialize(mapload)
+	. = ..()
+	pedometer = new(src, CALLBACK(src, PROC_REF(on_stepped)))
+
+/obj/item/reagent_containers/food/snacks/egg/watcher/Destroy(force)
+	. = ..()
+	QDEL_NULL(pedometer)
+
+/obj/item/reagent_containers/food/snacks/egg/watcher/examine(mob/user)
+	. = ..()
+	if(steps_travelled < (steps_to_hatch * WATCHER_EGG_ACTIVE_MOD))
+		. += "<span class='notice'>Something stirs listlessly inside.</span>"
+	else if(steps_travelled < steps_to_hatch * WATCHER_EGG_LIVELY_MOD)
+		. += "<span class='notice'>Something is moving actively inside.</span>"
+	else
+		. += "<span class='boldnotice'>It's jiggling wildly, it's about to hatch!</span>"
+
+
+
+/// Called when we are moved, whether inside an inventory or by ourself somehow
+/obj/item/reagent_containers/food/snacks/egg/watcher/proc/on_stepped(atom/movable/egg, atom/mover, atom/old_loc, direction)
+	var/new_loc = get_turf(egg)
+	if(isnull(new_loc) || new_loc == get_turf(old_loc))
+		return // Didn't actually go anywhere
+	steps_travelled++
+	if(steps_travelled == steps_to_hatch * WATCHER_EGG_ACTIVE_MOD)
+		jiggle()
+		return
+	if(steps_travelled < steps_to_hatch)
+		return
+	visible_message("<span class='boldnotice'>[src] splits and unfurls into a baby Watcher!</span>")
+	playsound(new_loc, 'sound/effects/splat.ogg', 50, TRUE)
+	new /obj/effect/decal/cleanable/greenglow(new_loc)
+	new /obj/item/watcher_hatchling(new_loc)
+	qdel(src)
+
+/// Animate the egg
+/obj/item/reagent_containers/food/snacks/egg/watcher/proc/jiggle()
+	var/animation = isturf(loc) ? rand(1, 3) : 1 // Pixel_x/y animations don't work in an inventory
+	switch(animation)
+		if(1)
+			animate(src, transform = transform.Scale(1.3), time = 1 SECONDS, easing = BOUNCE_EASING)
+			animate(transform = matrix(), time = 0.5 SECONDS, easing = SINE_EASING | EASE_IN)
+		if(2)
+			animate(src, pixel_y = 8, time = 0.5 SECONDS, easing = SINE_EASING | EASE_OUT)
+			animate(pixel_y = 0, time = 0.5 SECONDS, easing = SINE_EASING | EASE_IN)
+			animate(pixel_y = 4, time = 0.5 SECONDS, easing = SINE_EASING | EASE_OUT)
+			animate(pixel_y = 0, time = 0.5 SECONDS, easing = BOUNCE_EASING | EASE_IN)
+		if(3)
+			Shake(2, 0, 0.3 SECONDS)
+	var/next_jiggle = rand(5 SECONDS, 10 SECONDS) / (steps_travelled >= steps_to_hatch * WATCHER_EGG_LIVELY_MOD ? 2 : 1)
+	addtimer(CALLBACK(src, PROC_REF(jiggle)), next_jiggle, TIMER_DELETE_ME)
```

Comment:
Is this an infinite loop of self-calling?

Reply:
Correct, until it hatches